src/main/scala/supporters/functions/FunctionData.scala

```diff
@@ -186,120 +166,75 @@ class FunctionData(val programFunction: ast.Function,
-      val predicate = program.findPredicate(predAcc.predicateName)
-      val triggerFunction = predicateData(predicate).triggerFunction
+    if(translatedPres.nonEmpty)
+      combinedTerm = Implies(And(translatedPres), combinedTerm)
 
-      /* TODO: Don't use translatePrecondition - refactor expressionTranslator */
-      val args = (
-           expressionTranslator.getOrFail(locToSnap, predAcc, Seq(), sorts.Snap, Option.when(Verifier.config.enableDebugging())(PUnknown()))
-        +: expressionTranslator.translatePrecondition(program, predAcc.args, this))
+    combinedTerm = And(Equals(formalResult, App(functionSupporter.limitedVersion(function), arguments)), combinedTerm)
+    transformAllFunctionCalls(combinedTerm, functionSupporter.postconditionVersion)
+  }
 
-      val fapp = App(triggerFunction, args)
+  def transformAllFunctionCalls(term: Term, transformer: HeapDepFun => HeapDepFun): Term = {
+    val apps = term.deepCollect({ case app: App if app.applicable.isInstanceOf[HeapDepFun] &&  app.applicable.asInstanceOf[HeapDepFun] != functionSupporter.limitedVersion(app.applicable.asInstanceOf[HeapDepFun]) => app})
+    var i = 0
+    val resFunMap = apps.map(f => {
+      i += 1
+      f -> Var(SimpleIdentifier("res@" + f.applicable.id.name + "@" + i), f.sort, false)
+    })
 
-      predicate -> fapp
-    }))
-  }
+    var replacedTerms = term.replace(resFunMap.map(_._1), resFunMap.map(_._2))
 
-  lazy val optBody: Option[Term] = {
-    assert(phase == 2, s"Definitional axiom must be generated in phase 2, current phase is $phase")
+    resFunMap.foreach(v => {
+      val res = v._2
+      val args = v._1.args.map(_.replace(resFunMap.map(_._1), resFunMap.map(_._2)))
+      val origFun = v._1.applicable.asInstanceOf[HeapDepFun]
+      val limitedApp = App(functionSupporter.limitedVersion(origFun), args)
 
-    expressionTranslator.translate(program, programFunction, this)
-  }
+      val funToCall = transformer(origFun)
+      val app = App(funToCall, args.appended(res))
+      replacedTerms = Let(res, limitedApp, And(replacedTerms, app))
+    })
 
-  lazy val definitionalAxiom: Option[Term] = {
-    assert(phase == 2, s"Definitional axiom must be generated in phase 2, current phase is $phase")
-
-    optBody.map(translatedBody => {
-      val pre = preconditionFunctionApplication
-      val nestedDefinitionalAxioms = generateNestedDefinitionalAxioms
-      val body = And(nestedDefinitionalAxioms ++ List(Implies(pre, And(BuiltinEquals(functionApplication, translatedBody)))))
-      val funcAnn = programFunction.info.getUniqueInfo[ast.AnnotationInfo]
-      val actualPredicateTriggers = funcAnn match {
-        case Some(a) if a.values.contains("opaque") => Seq()
-        case _ => predicateTriggers.values.map(pt => Trigger(Seq(triggerFunctionApplication, pt)))
-      }
-      val allTriggers = (
-           Seq(Trigger(functionApplication)) ++ actualPredicateTriggers)
-
-      Forall(arguments, body, allTriggers)})
+    replacedTerms
```

Comment:
I don't think I understand what's going on here. You build up a big ``let res1 = f%limited(...) in let res2 = g%limited(...) in term`` expression for all non-limited function calls in term I guess? Isn't that equivalent to just replacing all calls with their limited version?

Reply:
cleaned it up such that it does not use let terms anymore